openff/toolkit/topology/molecule.py

```diff
@@ -1034,6 +1037,8 @@ def __init__(
                     msg += str(value_error)
                 raise ValueError(msg)
 
+        self._name = name
```

Comment:
Oh, wow, I now see how this is a disaster. Most of the code paths above that could have loaded a molecule have their own logic for setting that molecule's name, and this would unconditionally overwrite all of that. But to do anything else would require deciding, implementing, documenting, and testing how the priority works out in the event of each conflict.

I hate this initializer so much.

I've been going in circles for like an hour trying to figure out how to make this work and my train of thought is either fully derailed or this is actually hard, but I figure this will at least get partial credit for putting `name` in `Molecule.from_smiles` and I'm gonna redefine this PR to only do that and `from_inchi`. We can figure out how to handle this toxic waste pile of an `__init__` in a different PR.

Reply:
Agree, modulo any details I've forgotten since the last time I looked at this code.

In this PR I tried to set the object's name in as many code paths as possible with a little changes to the existing signatures as possible, i.e. some use kwargs and some don't. The major exceptions are `Molecule.from_file` since that umbrellas across so much behavior and some stuff like `.from_file_obj` because I wanted to avoid looking into it. We could just as easily split this into smaller, more targeted PRs for the various input paths. (I forget how many I changed and right now don't want to take the time to survey it.)

I'd be in support of a more substantial rewrite of the initializer and some of the related code paths, even if that meant some behavior changes. The pattern of passing a string to the initializer is possibly the source of most of the issues, given the expansive set of possible inputs it needs to reason through. It'd be as easy to claw back as a toy from a toothing toddler, though ...